gk/main.c

```diff
@@ -400,20 +491,48 @@ add_ggu_policy(struct ggu_policy *policy, struct gk_instance *instance)
 		/* Create a new flow entry. */
 		ret = rte_hash_add_key_with_hash(
 			instance->ip_flow_hash_table,
- 			(void *)&policy->flow, rss_hash_val);
+ 			&policy->flow, rss_hash_val);
 		if (ret < 0) {
 			RTE_LOG(ERR, HASH,
 				"The GK block failed to add new key to hash table!\n");
 			return;
 		}
 
+		fe_existing = false;
+
 		fe = &instance->ip_flow_entry_table[ret];
-		initialize_flow_entry(fe, &policy->flow);
+		/*
+		 * It only fills up GK_GRANTED and GK_DECLINED states,
+		 * so, it doesn't need to call initialize_flow_entry().
+		 */
+		rte_memcpy(&fe->flow, &policy->flow, sizeof(fe->flow));
 	} else
 		fe = &instance->ip_flow_entry_table[ret];
 
-	switch(policy->state) {
-	case GK_GRANTED:
+	switch (policy->state) {
+	case GK_GRANTED: {
+		int gt_id = -1;
+
+		/*
+		 * When the flow entry already exists,
+		 * the grantor ID may be already known
+		 * and save the lookup.
+		 */
+		if (fe_existing && (fe->state == GK_REQUEST
+				|| fe->state == GK_GRANTED)) {
+			if (fe->state == GK_REQUEST)
+				gt_id = fe->u.request.grantor_id;
+			else if (fe->state == GK_GRANTED)
+				gt_id = fe->u.granted.grantor_id;
```

Comment:
When the change I suggested in `reinitialize_flow_entry()`, the grantor id will always be known when `fe` already exists. Given that `reinitialize_flow_entry()` and here needs the code to get the grantor id from an existing `fe`, you should create an inline function to encapsulate this code. My suggestion to the name of this new function: `grantor_id_of_flow_entry()`.